### PR TITLE
Remove unecessary dependencies

### PR DESCRIPTION
**Pull request by Aschen**
## Purpose

This is a very nice project, I like everything that tries to make Javascript ecosystem more secure ;-)
And especially if the tool comes with a nice CLI and GUI :D 

The purpose of this PR is to try try to get ride of some of the dependencies used by `nsecure`.  

This is motivated by:
 - reduce total downloaded code
 - reduce potential attack surface

The following dependencies have been removed:
 - `filenamify`
 - `get-port`
 - `lodash.get`
 - `make-promise-safe`
 - `unique-slug`

(`httpie` has just been moved to `devDependencies`)

I think you could also get ride of these dependencies:
 -  `is-minified-code` the algorithm is pretty straightforward 
 - `itertools` fortunately the STD lib offer a wide range of functionalities on collection  
 - `lodash.clonedeep` you could simply use `JSON.parse(JSON.stringify(..))`` for most of your use cases (you may need to stop using symbol to declare class member thus)

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -143,7 +140,7 @@ function loadNsecureCache(defaultPayload = {}) {
         return JSON.parse(buf.toString());
     }
 
-    const payload = Object.assign({}, cloneDeep(defaultPayload), {
+    const payload = Object.assign({}, JSON.parse(JSON.stringify(defaultPayload)), {
```
Comment (Xstoudi) [inline src/utils.js:143]: Shouldn't do that. JSON.stringify comes with unexpected behaviors like below and is really slow.

```js
JSON.parse(JSON.stringify({ foo: Infinity}))
// {  key: null }
```

File: test/getTarballComposition.js [hunk 2]
```diff
@@ -25,5 +25,5 @@ test("should return the composition of a directory", async() => {
         size
     });
     expect(composition.files).toHaveLength(4);
-    expect(composition.files[0]).toMatch(/one(\/|\\)README/);
+    expect(composition.files.sort()[0]).toMatch(/one(\/|\\)README/);
```
Comment (Xstoudi) [inline test/getTarballComposition.js:28]: What's the point here?

File: src/utils.js [hunk 3]
```diff
@@ -175,7 +172,76 @@ function taggedString(chaines, ...cles) {
...
+
+// eslint-disable-next-line
+const FORBIDDEN_FILENAME_CHAR_REG = /[<>:"\/\\|?*\x00-\x1F]/g;
+
+function filenamify(filename, replacement = "!") {
+    return filename.replace(FORBIDDEN_FILENAME_CHAR_REG, replacement);
+}
+
+// from https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_get
+function get(obj, path, defaultValue) {
```
Comment (Xstoudi) [inline src/utils.js:195]: Seems slower than _.get. As only 'loadash.get' is required, I don't really think it's a big deal.

File: src/utils.js [hunk 4]
```diff
@@ -175,7 +172,76 @@ function taggedString(chaines, ...cles) {
...
+                tester.once("close", () => resolve(true)).close();
+            })
+            .listen(port);
+    });
+}
+
+async function getPort(portNumber) {
+    let idx = portNumber;
+
+    while (idx < 65535) {
```
Comment (Xstoudi) [inline src/utils.js:229]: Ports below 1024 arent's suitable for node-secure, should sanity check the `portNumber` variable.

**Comment by fraxken** [comment]
Hi,

First thanks for your contribution (I really appreciate.). To be honest your pull-request is a bit "too much" at the same time (there is idea that I like and some I don't.).

As the creator of Node-secure, I'm very aware of what kind of dependencies I use, and I'm currently working hard to update and remove legacy dependencies from packages like cacache, pacote etc. (which introduce a lot of indirect dependencies).



For example you want to remove "make-promises-safe" as it was "just" few line of codes to copy past... But no! Make-promises-safe is maintained by a Node.js core contributor who take care of testing and evolving the code to follow the Deprecation 18 (and others dep).

I guess uniqueSlug may be overkilled for what we do/want in Node-secure (Something like crypto.randomBytes should be able to achieve equivalent work with no need of new implementation.).

Notes:

- httpie is a production dep (used in httpServer.js).
- itertools is about methods for sync and async generators (not about array). But i guess 33KB is overkill for what i'm doing with it right now.
- the guy behind is-minified-code is cool and has already answered multiple of my questions when i needed to exchange on his algorithm.

Best Regards,
Thomas

File: src/utils.js (hunk 4 above)
Comment (targos) [reply src/utils.js:229]: can't we use port `0` instead and let the operating system choose an available port for us?

File: src/utils.js (hunk 1 above)
Comment (Aschen) [reply src/utils.js:143]: Oh I didn't know that one :+1:  I only known about key order instability.  

About the performances you are right, `cloneDeep` is slightly more performant on small object but when the object get bigger JSON.parse can be more effective (see [this explanation](https://www.youtube.com/watch?v=ff4fgQxPaO0):
```
clone deep x 329,985 ops/sec ±2.83% (81 runs sampled)
json x 217,681 ops/sec ±2.93% (76 runs sampled)
clone deep (big object) x 58,546 ops/sec ±2.29% (79 runs sampled)
json (big object) x 56,364 ops/sec ±2.29% (83 runs sampled)
```

File: src/utils.js (hunk 3 above)
Comment (Aschen) [reply src/utils.js:195]: Yes this method is 4x slower but still very fast
```
lodash get x 4,762,210 ops/sec ±1.67% (87 runs sampled)
jsgeton x 1,240,098 ops/sec ±1.72% (88 runs sampled)
```
IMHO in a non-critical section of code there is no incentive to choose lodash over this method.  

The point is still trying to reduce external dependencies. Even if `lodash.get` is small, it's still an unnecessary dependency

**Comment by Aschen** [comment]
Hi @fraxken,

I didn't wanted to force you or something, sorry if you feel it that way. It's just a small contribution to your project to try to reduce external dependency as much as I can.

I saw that you were also trying to reduce dependencies, which is good :+1:  Actually I know it's also an issue in many of our product and I will be very happy if someone come up with a PR for this :sweat_smile: 

I agree that some small dependencies can still be useful since their maintainers follow Node.js and v8 evolution ([this package](https://github.com/davidmarkclements/flatstr) for example had to follow many evolution of v8).

But I disagree with you on this point for `make-promise-safe` because IMHO the deprecation is here and it will not change in any way (apart from crashing Node.js in the next LTS) so IMHO it's safe to just attach a listener on `unhandledRejection` (see [file; history here](https://github.com/mcollina/make-promises-safe/commits/master/make-promises-safe.js))

About `unique-slug`, this package is the perfect example of unnecessary and pointless dependency:
 - it can be achieve with 10 lines of code
 - it's inefficient since a murmur hash has to be generated

Also IMHO you don't need to use the crypto API for a random slug. This API should be used only for cryptographic purpose like key or iv generation because it's very slow and also you system have a limited of pseudorandom bit to offer and other application may need it.

For `httpie`, I only noticed an usage in the`test/httpServer.js` file, that's why I put it in `devDependencies` but maybe I missed something :thinking: 

Finally about `itertools`, I tried to replace it and I almost succeed but then I figured out that I had to refactor more part of the code and I'm not really good with async generators so I didn't wanted to deep dive on this one.

File: src/utils.js (hunk 4 above)
Comment (Aschen) [reply src/utils.js:229]: I didn't even know that we can do that!  
As you want guys, the advantage of this method is that most of the time you can have a fixed time so your user wont get confused when they want to open the web app

**Comment by targos** [comment]
It doesn't seem right that the entire package-lock is modified by this PR. GitHub's UI doesn't want to show the diff so I can't see what happened.

**Comment by Aschen** [comment]
> It doesn't seem right that the entire package-lock is modified by this PR. GitHub's UI doesn't want to show the diff so I can't see what happened.

@targos It's look like git wasn't able to properly diff the two different version of the `package-lock.json` but I just ran `npm install` after removing dependencies.

**Comment by fraxken** [comment]
Hi,

Sorry for the late response (I didn't really take the time to go back to this PR.. sorry). Because of this PR and the reviews we landed a lot of individual commits on the master branch so it was very useful:

- https://github.com/ES-Community/nsecure/commit/677404103b3f772937d541989da5ab1032049221
- https://github.com/ES-Community/nsecure/commit/27e3cbb78fd57375d55f33f400e7648e8ad210c1

(And surely others that i missed).

However, as discussed from the beginning, I did not necessarily agree with all the changes. In the future I will surely ask for PR that only remove one dependency at a time (Otherwise it's really complicated to discuss the matter with others and decide  whether we want to implement the change or not).

Best Regards,
Thomas

**Pull request closed without merging**